### PR TITLE
CoreDataHelper: Extensions are way cooler

### PR DESCRIPTION
**Pull request by jleandroperez**
### Description:
In this PR we're converting `CoreDataHelper` into a `NSManagedObjectContext` extension, and implementing a mechanism that gives all NSManagedObject subclasses a default `class var entityName` (which can be overriden).

This was, initially, my goal. But Swift's compiler segmentation faults forced us to implement the helpers as a Struct.

This is Mark II, with a (hopefully) way more stable compiler.

### To test:
- Run the unit tests, verify that everything is green
- Smoke test the App (specially Notifications!)

Needs review: @astralbodies 
Thanks in advance sir!

### CODE REVIEW

**Comment by astralbodies** [comment]
I'll have to review this Monday @jleandroperez - if I get a chance over the weekend I shall do so. 😄

**Comment by jleandroperez** [comment]
@astralbodies this one is low priority / tech debt. It definitely can wait, thank you sir!!

P.s.: Have fun!!!

File: WordPress/Classes/Utility/CoreDataHelper.swift [hunk 1]
```diff
@@ -1,47 +1,44 @@
...
     ///
     /// - Parameter predicate: Defines the conditions that any given object should meet. Optional.
     ///
-    func countObjects(matchingPredicate predicate: NSPredicate? = nil) -> Int {
-        let request = newFetchRequest()
+    func countObjects<T: NSManagedObject>(ofType type: T.Type, matching predicate: NSPredicate? = nil) -> Int {
+        let request = newFetchRequest(forType: type)
         request.predicate = predicate
         request.includesSubentities = false
         request.predicate = predicate
```
Comment (astralbodies) [inline WordPress/Classes/Utility/CoreDataHelper.swift]: Not something you introduced in this PR but should get fixed - setting predicate twice.

File: WordPress/Classes/Utility/CoreDataHelper.swift [hunk 2]
```diff
@@ -1,47 +1,44 @@
...
-    ///
-    let context: NSManagedObjectContext
-
+extension NSManagedObjectContext {
 
     /// Returns a new FetchRequest for the associated Entity
     ///
-    func newFetchRequest() -> NSFetchRequest<NSFetchRequestResult> {
-        return NSFetchRequest(entityName: T.entityName)
+    func newFetchRequest<T: NSManagedObject>(forType type: T.Type) -> NSFetchRequest<NSFetchRequestResult> {
```
Comment (astralbodies) [inline WordPress/Classes/Utility/CoreDataHelper.swift]: This probably shouldn't be in the extension of `NSManagedObjectContext` as it's not related to a context in any way until execution. `NSManagedObject` already has a `fetchRequest` class function on it for doing exactly this. I say delete and refactor as such: 

```swift
let request = T.fetchRequest()
```

File: WordPress/Classes/Utility/CoreDataHelper.swift (hunk 1 above)
Comment (jleandroperez) [inline WordPress/Classes/Utility/CoreDataHelper.swift]: Huh thank you!!

File: WordPress/Classes/Utility/CoreDataHelper.swift (hunk 2 above)
Comment (jleandroperez) [inline WordPress/Classes/Utility/CoreDataHelper.swift]: Good call! This was initially implemented when we had iOS 9 support on the table, and definitely forgot about the iOS 10 helper. Replacing, thanks for spotting this!!

**Comment by jleandroperez** [comment]
`NSMO`'s subclasses `entity()` helper doesn't work properly in Unit Testing targets. Definitely related to some inner workings (main suspect is: different Bundle!).

I've pushed a compromised solution in which we use the `entityName` set in the entity, *OR* the **className**, whenever the entity name cannot be determined.

@astralbodies ready for another round!! thanks in advance!!

**Comment by jleandroperez** [comment]
Thanks for the review sir!!

**Pull request merged**